include/kokkos-utils/callbacks/TimerListener.hpp

```diff
@@ -0,0 +1,169 @@
+#ifndef KOKKOS_UTILS_CALLBACKS__TIMERLISTENER_HPP
+#define KOKKOS_UTILS_CALLBACKS__TIMERLISTENER_HPP
+
+#include "Kokkos_NumericTraits.hpp"
+
+#include "kokkos-utils/callbacks/Listener.hpp"
+#include "kokkos-utils/callbacks/Matcher.hpp"
+#include "kokkos-utils/concepts/ExecutionSpace.hpp"
+#include "kokkos-utils/timer/Timer.hpp"
+
+namespace Kokkos::utils::callbacks
+{
+
+template <typename...>
+class TimerListener;
+
+/**
+ * @brief Listener that starts a timer when a begin event is received and stops it
+ *        when a corresponding end event is received.
+ *
+ * The start event is selected by requiring it to be of type @p BeginEventType and
+ * to be a match for the passed matcher. The stop event must be of type @p EndEventType
+ * and have the same ID as the begin event.
+ *
+ * The underlying timer @ref timer is a @ref Kokkos::utils::timer::Timer, instantiated
+ * for the execution space @p Exec. This class stores the execution space
+ * instance passed to the constructor, and uses it as the one passed to @ref timer.
+ *
+ * This class is well-suited for timing the execution of a parallel workload, such as a
+ * @c Kokkos parallel for, parallel reduce, parallel scan, or deep copy operation, to which
+ * an execution space instance is passed. In such a situation, the execution space instance
+ * passed to this class's constructor should be the same as the one passed to the workload.
+ */
+template <
+    Kokkos::utils::callbacks::BeginEvent BeginEventType, Kokkos::utils::callbacks::EndEvent EndEventType,
+    Kokkos::utils::callbacks::Matcher<Kokkos::Impl::type_list<BeginEventType>> MatcherType,
+    Kokkos::utils::concepts::ExecutionSpace Exec
+>
+class TimerListener<BeginEventType, EndEventType, MatcherType, Exec>
+{
+public:
+    static constexpr uint64_t invalid_event_id = Kokkos::Experimental::finite_max_v<uint64_t>;
+
+public:
+    template <typename T> requires (std::same_as<std::remove_cvref_t<T>, MatcherType>)
+    TimerListener(Exec exec_, T&& matcher_)
+      : exec(std::move(exec_)), matcher(std::forward<T>(matcher_)) {}
+
+    /**
+     * @todo Consider optimizing the implementation to avoid repeating the regex many
+     *       times after the timer is already closed.
+     *
+     * @todo Allow the possibility to combine multiple matchers with logical operators, so that
+     *       we can match e.g. a begin event with a specific name within a specific section.
+     */
+    void operator()(const BeginEventType& event) {
+        if (matcher(event)) {
+            if (this->connected() || this->closed()) Kokkos::abort("Timer cannot match a begin event twice.");
```

Comment:
I guess you can get a substantial reduction of the overhead of this timer by skipping any event that is not launched on `exec`. An integer comparison is much cheaper than a regex matching.

If you do it, please also:
- precompute the `dev_id` of `exec` once and forall ?
- add a test that launches *e.g.* two parallel reduce with the same label (so both would match) but one is on `exec` and the other one is on `exec_other`. You should only match one.

```suggestion
    void operator()(const BeginEventType& event)
    {
        if(event.dev_id != Kokkos::Tools::Experimental::device_id(exec)) return;

        if (matcher(event)) {
            if (this->connected() || this->closed()) Kokkos::abort("Timer cannot match a begin event twice.");
```

Reply:
OK. Good suggestion indeed! I've modified it and added tests. The test of the case where we submit on one exec, but the TimerListener is on another exec didn't pass. The reason appears to be that we test only on `OpenMP`, and, on `OpenMP`, it appears that the `device_id` is the same for each instance:

https://github.com/kokkos/kokkos/blob/88485ca664e39a93725b44f99d3c4b03fcf701e7/core/src/OpenMP/Kokkos_OpenMP.hpp#L142

And so the `TimerListener` accepts the event, even if it's on another `exec`.

Reply:
Simply adapt the test by adding something like:
```c++
ASSERT_THAT(event accepted, only if exec A dev id equals exec B dev id);
```

It makes sense that the device ID is the same for several `OpenMP` instances, because they aren't memory pools and are on the same CPU (?).

Reply:
And we should probably add a GPU runner.

Reply:
OK. Added such a test. Let's see what the new GPU runner does :)

Reply:
Solved indeed!